### PR TITLE
[WIP] Make ApplicationLifetime aware of OnFrameworkInitializationCompleted

### LINKED ISSUE #3224: FontManager.Default is always empty for desktop application
FontManager.Default is always empty fontmanager for desktop application.

repro:

Start ControlCatalog Dektop from master and in some Window constructor type to check:
`var fm = FontManager.Default;`

FontManager.Default is EmptyFontManager for both Skia and Direct2D

Initialization somewhere is not in the right order.
I believe it's the static ctor of ClassicDesktopStyleApplicationLifetime.

### PR DESCRIPTION
**Pull request by Gillibald**
## What does the pull request do?
It allows a lifetime object to know when the framework has beed initialized to do work that is only safe to perform when all services are available.

## What is the current behavior?
Currently it is not possible for a lifetime object to know when the framework has been initialized. This causes some issues with the ClassicDesktopStyleApplicationLifetime. It subscribes to window events in its static ctor when not all services are available.

## What is the updated/expected behavior with this PR?
`IApplicationLifetime` has a new method `OnFrameworkInitializationCompleted()` that is called when the framwork is initialized.

## How was the solution implemented (if it's not obvious)?
`IApplicationLifetime.OnFrameworkInitializationCompleted()` was introduced and is called by the owning application when its `OnFrameworkInitializationCompleted()` is called.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Fixed issues
Fixes: #3224

### CODE REVIEW

File: src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs [hunk 1]
```diff
@@ -96,21 +79,45 @@ public void Shutdown(int exitCode = 0)
             }
         }
         
-        
         public int Start(string[] args)
         {
+            _windowOpenedDisposable = Window.WindowOpenedEvent.AddClassHandler(typeof(Window), OnWindowOpened);
```
Comment (donandren) [inline src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs]: may be single  CompositeDisposable should make it bit easy to handle the dispose

File: src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs [hunk 2]
```diff
@@ -96,21 +79,45 @@ public void Shutdown(int exitCode = 0)
...
+
+            HandleWindowClosed((Window)sender);
+        }
+
+        private void OnWindowOpened(object sender, RoutedEventArgs e)
+        {
+            _windows.Add((Window)sender);
+        }
+
         public void Dispose()
```
Comment (donandren) [inline src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs]: not sure dispose is ever called, but that shouldn't be a problem

**Review by donandren — APPROVED** [review]
this should fix the issue with Default FontManager

File: src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs (hunk 1 above)
Comment (Gillibald) [reply src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs]: Y we can combine them. Will make that change tomorrow or feel free to do it yourself.

**Comment by Gillibald** [comment]
Will fix the tests tomorrow. They need a call to lifetime.Start

**Comment by kekekeks** [comment]
Seems to be messing up the original logic with exactly one set of handler with exactly one active lifetime. 

Can we introduce `EnsureInitialized` static method instead?

**Review by kekekeks — CHANGES_REQUESTED** [review]
Seems to be messing up the original logic with exactly one set of handler with exactly one active lifetime. 

Can we introduce `EnsureInitialized` static method and call it from non-static `ClassicDesktopStyleApplicationLifetime` constructor instead?

**Comment by Gillibald** [comment]
There can't be multiple lifetime instances. Touching anything UI related before Start has run breaks. I don't see how we can solve this problem differently.

**Comment by Gillibald** [comment]
We can construct the lifetime after everything was initialized. Ghat would solve this too and we could keep the current version.

**Comment by kekekeks** [comment]
>There can't be multiple lifetime instances

Our tests re-create lifetimes on each run.

**Comment by kekekeks** [comment]
BTW, this isn't exactly right approach:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/Media/FontManager.cs#L62

**Comment by kekekeks** [comment]
https://github.com/AvaloniaUI/Avalonia/issues/1013

**Comment by Gillibald** [comment]
I asked you in Gitter if it is fine to have a global font manager and you were fine with it. We can bind this to the top level but if we do that static property definitions like `FontFamily` would still cause problems.

**Comment by Gillibald** [comment]
Closing this in favor of a more robust fix

**Comment by donandren** [comment]
I don't think we should close the pr @Gillibald @kekekeks , the `ClassicDesktopStyleApplicationLifetime` static ctor calling initializations in avalonia (from Window static ctor) before platform is setup could bring another issue in the future, so this is a problem not only related to `FontManager`

**Comment by Gillibald** [comment]
> Your PR wouldn't work properly anyway, BTW
Since Window instances could be created and opened in OnFrameworkInitializationCompleted
which happens before Start
What we could do is some kind of extra initialization step
we have AfterPlatformServicesSetup extension point
which seems to be a good place for actually initializing app lifetimes

**Comment by kekekeks** [comment]
>the ClassicDesktopStyleApplicationLifetime static ctor calling initializations in avalonia (from Window static ctor) before platform is setup could bring another issue in the future

We have several static constructors relying on AvaloniaLocator to be ready. That is ultimately the root of the issue here. Several parts of the code base (like FontManager in question) even have `beforefieldinit` static constructors, which allows (according to [ECMA-355](https://www.ecma-international.org/publications/standards/Ecma-335.htm) I.8.9.5) the runtime to call them at *any* point of time, even before the `Main` method

That makes the whole thing fragile and unreliable. Static constructors should not under any circumstances touch AvaloniaLocator. If they do, it should be considered to be a bug and get fixed sooner or later.


Accessing AvaloniaObject metadata should *not* rely on platform-specific features. Registering properties, adding property owners, etc, should *not* break things if done before platform initialization.

**Comment by kekekeks** [comment]
Regarding the current PR. At the time when `Start` method is called, the user code in `Application.OnFrameworkInitializationCompleted` is already executed. So there might already be some opened windows that won't be tracked by the lifetime if it subscribes to events in `Start`

File: src/Avalonia.Controls/AppBuilderBase.cs [hunk 3]
```diff
@@ -163,9 +163,10 @@ public TAppBuilder SetupWithLifetime(IApplicationLifetime lifetime)
         {
             _lifetime = lifetime;
             Setup();
+            lifetime.OnSetupCompleted();
```
Comment (kekekeks) [inline src/Avalonia.Controls/AppBuilderBase.cs]: This is called after `Application.OnFrameworkInitializationCompleted`

**Comment by Gillibald** [comment]
`Application.OnFrameworkInitializationCompleted` now calls `IApplicationLifetime.OnFrameworkInitializationCompleted`

ClassicDesktopStyleApplicationLifetime was never disposed so I remove IDisposable

**Comment by kekekeks** [comment]
>never disposed

>10 `using` blocks in https://github.com/AvaloniaUI/Avalonia/blob/master/tests/Avalonia.Controls.UnitTests/DesktopStyleApplicationLifetimeTests.cs#L19

**Comment by Gillibald** [comment]
Only tests where using IDisposable that seems to be a bit broken. A real application never disposes the lifetime.

**Comment by kekekeks** [comment]
App might try to restart itself with a different lifetime.

**Comment by Gillibald** [comment]
Then IApplicationLifetime should implement IDisposable to signal the lifetime should end.

**Comment by grokys** [comment]
@Gillibald could you enter a description for this PR? It's important to help understand 5 years down the line why something was done.

**Comment by kekekeks** [comment]
>Then IApplicationLifetime should implement IDisposable to signal the lifetime it should end.

That's not possible on certain platforms.

**Comment by Gillibald** [comment]
How else would you have a clean end of a lifetime if you were to replace it? Allocation and ApplicationLifetime are so tightly coupled that it should be just one component. I don't see how one can exist without the other. Currently one can just replace the lifetime and nothing reacts to it.

**Comment by Gillibald** [comment]
Lets revisit this in the near future

**Pull request closed without merging**